app/models/ability.rb

```diff
@@ -0,0 +1,31 @@
+# frozen_string_literal: true
+class Ability
+  include Hydra::Ability
+
+  def custom_permissions
+    can :manage, Valkyrie::Model if current_user.admin?
+    alias_action :file_manager, to: :update
+  end
+
+  def read_permissions
+    super
+    can :read, Valkyrie::Model do |obj|
+      valkyrie_test_read(obj) || valkyrie_test_edit(obj)
+    end
+  end
+
+  def edit_permissions
+    super
+    can [:edit, :update, :destroy], Valkyrie::Model do |obj|
+      valkyrie_test_edit(obj)
+    end
+  end
+
+  def valkyrie_test_read(obj)
+    !(user_groups & obj.read_groups).empty? || obj.read_users.include?(current_user.user_key)
```

Comment:
I look at this kind of expression as a tax on maintainers. It's tucked away in a named method, but then it's hiding two more useful, easily-named primitives. I don't see this concern as "DRY at all costs" as much as getting the level abstraction tuned in.

Here, I see this method's purpose as expressing "user can read object" as having group-read or individual-read permission, but it's fiddling with set and boolean operations that require human parsing to impute that meaning.